runtime/src/main/java/org/corfudb/runtime/view/replication/ChainReplicationProtocol.java

```diff
@@ -68,83 +71,107 @@ public ILogData peek(RuntimeLayout runtimeLayout, long globalAddress) {
+            String logServer = logServers.get(logServers.size() - 1);
+            serverAddressMap.compute(logServer, (server, list) -> {
+                if (list == null) {
+                    // The additional wrapping is needed as we expect a resizable list
+                    return new ArrayList<>(Collections.singletonList(address));
+                }
+                list.add(address);
+                return list;
+            });
+        }
 
+        // Send read requests to log unit servers in parallel
+        List<CompletableFuture<ReadResponse>> futures = serverAddressMap.entrySet().stream()
+                .map(entry -> runtimeLayout.getLogUnitClient(entry.getKey()).readAll(entry.getValue()))
+                .collect(Collectors.toList());
 
-    @Override
-    public Map<Long, ILogData> readRange(RuntimeLayout runtimeLayout, Set<Long> globalAddresses) {
-        Range<Long> range = Range.encloseAll(globalAddresses);
-        return readRange(runtimeLayout, range, true);
+        // Merge the read responses from different log unit servers
+        Map<Long, LogData> readResult = futures.stream()
+                .map(future -> CFUtils.getUninterruptibly(future).getAddresses())
+                .reduce(new HashMap<>(), (map1, map2) -> {
+                    map1.putAll(map2);
+                    return map1;
+                });
+
+        return waitOrHoleFill(runtimeLayout, readResult, waitForWrite);
     }
 
-    /**
-     * Reads a range of addresses from the Chain of log unit servers. This method optimizes for the time to wait to
-     * hole fill in case an empty address is encountered.
-     * If the waitForWrite flag is set to true, when an empty address is encountered, it waits for the hole to be
-     * filled. All subsequent empty addresses are hole filled directly and the reader does not wait.
-     * In case the flag is set to false, none of the reads wait for write completion and the empty address is hole
-     * filled right away.
-     *
-     * @param runtimeLayout Runtime layout.
-     * @param range         Range of addresses to read.
-     * @param waitForWrite  Flag whether wait for write is required or hole fill directly.
-     * @return Map of read addresses.
-     */
-    public Map<Long, ILogData> readRange(RuntimeLayout runtimeLayout, Range<Long> range, boolean waitForWrite) {
-        long startAddress = range.lowerEndpoint();
-        long endAddress = range.upperEndpoint();
-        int numUnits = runtimeLayout.getLayout().getSegmentLength(startAddress);
-        log.trace("readRange[{}-{}]: chain {}/{}", startAddress, endAddress, numUnits, numUnits);
+    private Map<Long, ILogData> waitOrHoleFill(RuntimeLayout runtimeLayout,
+                                               Map<Long, LogData> readResult,
+                                               boolean waitForWrite) {
+        Map<Long, ILogData> returnResult = new TreeMap<>(readResult);
```

Comment:
Is there any instance where we need to do a multi read, but we don't want order over the addresses? Can't we just use A TreeMap instead of a Map since it seems like we always need order.

Reply:
Composed ordering in AddressSpaceView.